### DISCUSSION TITLE
Why strings are allowed in body if they are annotated as floats?

### DISCUSSION THREAD

**Discussion by blismatic** [fastapi/fastapi#12007] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# main.py
from typing import Annotated

from fastapi import FastAPI, Body
from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    name: str
    price: float
    
@app.put("/items/{item_id}")
def update_item(item_id: int, item: Annotated[Item, Body()]):
    result = {"item_id": item_id, "item": item}
    return result

# client.py
import requests

my_item = {
    "name": "Chicken",
    "price": "12.99"
}

response = requests.put("http://127.0.0.1:8000/items/1", json=my_item)
print(response.json())
```

### Description

Hello all, I am pretty new to FastAPI and have been reading the Tutorial - User Guide docs, but I came across a question that I couldn't seem to find an answer to. In the above example, I am defining that the `Item` class' `price` attribute needs to be of type `float`. However, in the actual PUT request that I am making, I am sending the `price` attribute as a `str`, `"12.99"`, and nothing is failing.

I understand that Pydantic is automatically converting the `str` `"12.99"` into the `float` `12.99` that my `update_item()` function is expecting. My question is, is this standard practice? Should the API response not return an error saying that the user sent a string, when it was expecting a float?

I first came across this question when reading this particular page of the docs: [https://fastapi.tiangolo.com/tutorial/schema-extra-example/#body-with-multiple-examples](https://fastapi.tiangolo.com/tutorial/schema-extra-example/#body-with-multiple-examples) It shows that it will accept the second example along with the first, but to me I (naively) assumed that it would/should reasonably fail just like the third example, since `"35.4"` is a string, even though I understand that it _could_ be converted to a float. It feels like that would be bad practice to allow a user of the API to send the "wrong" datatype when compared to what the schema in my `/docs` says: 
![image](https://github.com/user-attachments/assets/2df9610c-070a-49f1-ab65-a3d3e01af9fb)

### Operating System

Windows

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

Python 3.11.1

**Comment by JavierSanchezCastro** [comment, marked as answer]
You can make the field strict like that: 

```python3
from typing import Annotated
from fastapi import FastAPI, Body
from pydantic import BaseModel, Field

app = FastAPI()

class Item(BaseModel):
    name: str
    price: float = Field(strict=True)
    
@app.put("/items/{item_id}")
def update_item(item_id: int, item: Annotated[Item, Body()]):
    result = {"item_id": item_id, "item": item}
    return result
```
![imagen](https://github.com/user-attachments/assets/393313d0-bef0-4864-ba8e-20c23aa0311e)

It's more Pydantic side, you can also apply the strict mode for all fields using ConfigDict: https://docs.pydantic.dev/latest/concepts/strict_mode/#strict-mode-with-configdict

**Reply by blismatic** [reply]
Thanks for the response. Since this is the default behavior (converting types when possible), is this the best practice in most cases? Or should I be using `Field(strict=True)` in most places?

I haven't come across this yet in the FastAPI docs I have read so far, so I can only assume that this default behavior is standard practice... it just seems so strange to me. Perhaps I just need to accept it and stop worrying 😄

**Reply by JavierSanchezCastro** [reply]
Well, it depends on your needs and application, if you want to receive strict parameters, then you should use
`model_config = ConfigDict(strict=True)` for each BaseModel you create.

In my opinion, it's not something to worry about. Pydantic is going to manage for you those types of situations. But if, for whatever reason, you need that, you can create a BaseModelStrict like this and apply it to all your models:

```python3
from typing import Annotated
from fastapi import FastAPI, Body
from pydantic import BaseModel, ConfigDict

app = FastAPI()

class BaseModelStrict(BaseModel):
    model_config = ConfigDict(strict=True)

class Item(BaseModelStrict):
    name: str
    price: float
    
@app.put("/items/{item_id}")
def update_item(item_id: int, item: Annotated[Item, Body()]):
    result = {"item_id": item_id, "item": item}
    return result
```

**Reply by blismatic** [reply]
Thanks for the advice! Marking as answered. I really appreciate the help.
